### PR TITLE
Start bringing in matrixcn overall options and sparse matrix multiplier code

### LINKED ISSUE #987: Anomaly forcings
### Brief summary of bug

CESM2(CAM6) SSP scenarios that were run for CMIP6 scenarios has an aerosol emission bug that causes regional changes in climate (especially land temperatures over Asia).  I'm not sure how widely distributed the new anomaly forcings have been distributed, but we should likely be cautious using or sharing the new CMIP6 anomaly forcings for SSPs. The CESM2(WACCM) simulations are fine.

### General bug information

This bug would cause regional climate changes. New CESM2(CAM6) simulations are being run.

### Details of bug

In CESM2(CAM6) the anthropogenic and biomass burning Secondary Organic Aerosol (SOA) emissions had been set to zero starting in 2015, while the biogenic emissions were fine.

### Important details of your setup / configuration so we can reproduce the bug

Regional climate changes caused by the bug

### Important output or errors that show the problem

Gokhan has decided to "rerun our CESM2(CAM6) SSP Tier 1 simulations as soon as possible. We will also retract any impacted data sets from the ESGF site".

### LINKED ISSUE #1300: Update Anomaly forcing data and contrib script
@swensosc has updates to the ssp_anomaly_forcing_smooth in tools/contrib that also processes winds among other changes. The current script was used to create the data under $DIN_LOC_ROOT/tm/datm7/anomaly_forcing/CMIP6-SSP*/ directories. Also the simulations pointed to when that data was created has been superseded by new simulations because the emissions were wrong in the first set.

@danicalombardozzi

### LINKED ISSUE #1396: Add test for Anomaly forcing and rename test for  bias correction
Right now our test that's labeled "af" for anomaly forcing is actually a bias correction test (test mod af_bias_v7). This should be changed to be labeled "bc" for bias correction. And a new test added that's actually testing the anomaly forcing.

### LINKED ISSUE #1471: Add support for using "black" as a python formatter to our python coding process
We'd like to transition to using "black" as a python code formatter to our python coding process. Using it as a formatter will make sure we comply with PEP8, and keep our code in a consistent format. We talked about this in our CSEG meeting and CAM and others thought that there was a need for such a tool. pylint is still worthwhile in checking for issues, but we need a separate tool for formatting the appearance of code. Having a tool to do the formatting means we don't have to waste time working on appearance and can just let the formatter take care of consistent look and feel. We want to apply to this to the python modules under our "python" directory in CTSM. We won't require it of other python code, especially right now, and especially don't think the code in "tools/contrib" needs to worry about this.

There are two steps we think are important to accomplish this.

- Add a target to the python directory Makefile for "black" to run the python codes through black.
- Add a github action to check that python code has been run through black and if not trigger an error.

The first one should be straight forward as it will be similar to the pylint check in the python directory.

The second we aren't quite sure how it will work. What we'd like it to do is to prevent a PR from being approved if the "black check" operation isn't clear. This would be similar to the checkbox PR checker we already have in place. We think looking at how the action is done for the checkbox checker should help in figuring it out. Documentation on github for actions should also help. And black has some advice on github actions here...

https://black.readthedocs.io/en/stable/integrations/github_actions.html

I also see a lot of help on this from doing a google search for this, so there's likely many different examples of how to do this and we just need to pick the one for the way we do it.

@billsacks also suggests checking with Jesse in CAM and Jason Boutte from LBL.

### LINKED ISSUE #1755: Handle python environments for CTSM tools...
We had a subgroup to talk about how to setup python environments for our CTSM tools. @negin513 did some important work on establishing this as a need for both CSEG and CTSM. The group included: @wwieder @negin513 @adrifoster and @briandobbins as well as myself. This group and others feel free to comment and update this.

We talked about how to set this up and this is our thinking:

- We'll use conda for this (this is installed on both Cheyenne [with a module load] and is by default available on CGD machines
- Add manage_python_environment bash script to do the basic conda install at the top level (a user can do this by hand if they want to, the script will just do a few conda commands)
- Add a note on this to the relevant README files
- Use the YAML format for the file (allows you to set channels, and can allow pip specific commands as well)
- Have a basic file that sets the environment that will work for all CTSM tools
- Also have some more advanced environments for testing so we can figure out how advanced our python tools will work
- Put the yaml environment files under the python directory
- This same environment should work for the contrib tools as well as the supported python packages under the python directory if at all possible (we should try to NOT require a different environment for the contrib tools from the python packages).
- Some python packages should check the specific versions of python and modules explicitly
- Some python packages should do a "try, except" statement to give better error checking on if the python environment is correct.

Some thoughts I have that we didn't discuss, that I'd like to hear others opinion on:
- I think we should try to list the minimum set of things required to be set (rather than the long list of all dependencies)
- The environments should use explicit versions or an allowed range of versions, but not an open ">" (since in time there will be upper bound, so add a upper bound with "<", but add a comment about it)
- The python version should be explicit and set to the version available on CGD machines (right now this is restricted for the conda install in CGD) (setting the version to what's on CGD will allow us to work on both cheyenne and CGD machines)

### LINKED ISSUE #1795: SMS_D.f45_f45_mg37.I2000Clm50FatesRs.cheyenne_gnu.clm-FatesColdDefDryDep fails in ctsm5.1.dev099
### Brief summary of bug

The test SMS_D.f45_f45_mg37.I2000Clm50FatesRs.cheyenne_gnu.clm-FatesColdDefDryDep fails in ctsm5.1.dev099
it should be marked as an expected fail.

### General bug information

**CTSM version you are using:** ctsm5.1.dev099

**Does this bug cause significantly incorrect results in the model's science?** No

**Configurations affected:** Drydep and FATES without FATES-SP mode

### Details of bug

This is essentially #1044. It fails in trying to setup the namelist as the build-namelist dies when dry-dep is on without FATES-SP mode.

### PR DESCRIPTION
**Pull request by ekluzek**
### Description of changes

Start bringing in the use_matrixcn and use_soil_matrixcn control options for the Carbon/Nitrogen BGC model using
the matrix solution methodology developed by Yiqi Luo.

### Specific notes

Contributors other than yourself, if any:
   Yiqi Luo EcoLab members, Drs. Xingjie Lu, Yuanyuan Huang and Zhengguang Du, at Northern Arizona University

CTSM Issues Fixed (include github issue #):
  Fixes #1300
  Fixes #1396
  Fixes #987
  Fixes #1755
  Fixes #1795
  Fixes #1471

Are answers expected to change (and if so in what way)? No

Any User Interface Changes (namelist or namelist defaults changes)? No

Testing performed, if any: regular

### CODE REVIEW

**Comment by ekluzek** [comment]
This has passed testing on both cheyenne and izumi.

**Comment by ekluzek** [comment]
@glemieux and did a short code review over this. There was a couple suggestions that @glemieux had that I'll put in place. @glemieux will also add his review notes later.

- [ ] Make sure there is a comment for each matrixcn code section
- [ ] Some of those should have a bit about why they are empty
- [ ] Might add a comment about some of the use statements that are added in, but aren't really used (so someone doesn't remove it before the rest of the code comes in)
- [ ] Change the name of SPMod.F90 to CamelCase SparseMatrixMultiplyMod.F90 so its more descriptive

We also talked about the fact that we are bringing in custom Matrix handling code, and we wonder if general LAPACK code could be used to do the same thing. But, that will likely need to be a future refactoring.

**Review by glemieux — APPROVED** [review]
Looks good.  One question and minor fix to comments.

File: src/biogeochem/CNC14DecayMod.F90 [hunk 1]
```diff
@@ -120,26 +133,34 @@ subroutine C14Decay( bounds, num_soilc, filter_soilc, num_soilp, filter_soilp, &
...
+            livecrootc_storage(p) = livecrootc_storage(p)  * (1._r8 - decay_const * dt)
+            livecrootc_xfer(p)    = livecrootc_xfer(p)     * (1._r8 - decay_const * dt)
+            livestemc(p)          = livestemc(p)           * (1._r8 - decay_const * dt)
+            livestemc_storage(p)  = livestemc_storage(p)   * (1._r8 - decay_const * dt)
+            livestemc_xfer(p)     = livestemc_xfer(p)      * (1._r8 - decay_const * dt)
+         else
+            ! Each of these MUST correspond to the code above. Any changes in
+            ! code above need to apply here as well
+         end if
          gresp_storage(p)      = gresp_storage(p)       * (1._r8 - decay_const * dt)
```
[the remark this answers is not in the document]
Comment (ekluzek) [reply src/biogeochem/CNC14DecayMod.F90:162]: Yes.

File: src/main/filterMod.F90 [hunk 2]
```diff
@@ -96,6 +96,12 @@ module filterMod
      integer, pointer :: do_smb_c(:)     ! glacier+bareland SMB calculations-on filter (cols)
      integer :: num_do_smb_c             ! number of columns in glacier+bareland SMB mec filter
 
+     integer, pointer :: actfirec(:)     ! glacier+bareland SMB calculations-on filter (cols)
+     integer :: num_actfirec             ! number of columns in glacier+bareland SMB mec filter         
+
+     integer, pointer :: actfirep(:)     ! glacier+bareland SMB calculations-on filter (cols)
+     integer :: num_actfirep             ! number of columns in glacier+bareland SMB mec filter         
+
   end type clumpfilter
```
[the remark this answers is not in the document]
Comment (ekluzek) [reply src/main/filterMod.F90]: Thanks for pointing this out. I've updated, and also changed notes that refer to pfts to patches.

**Comment by ekluzek** [comment]
Testing on both cheyenne and izumi are as expected.

**Pull request merged**